app/shared/app-data/src/commonMain/kotlin/domain/torrent/TorrentServiceConnection.kt

```diff
@@ -0,0 +1,182 @@
+ *
+ * 此接口仅负责服务与 APP 之间的通信, 不负责服务的启动和终止.
+ */
+interface TorrentServiceConnection<T : Any> {
+    /**
+     * 当前服务是否已连接, 只有在已连接的状态才能获取通信接口.
+     *
+     * 若变为 `false`, 则服务通信接口将变得不可用, 可能需要实现类重新启动服务.
+     */
+    val connected: StateFlow<Boolean>
+
+    /**
+     * 获取通信接口. 如果服务未连接, 则会挂起直到服务连接成功.
+     *
+     * 这个函数是线程安全的.
+     */
+    suspend fun getBinder(): T
+}
+
+/**
+ * 通过 [Lifecycle] 来约束与服务的连接状态, 保证了:
+ * - 在 [RESUMED][Lifecycle.State.RESUMED] 状态下, 根据[文档](https://developer.android.com/guide/components/activities/activity-lifecycle#onresume), APP 被视为在前台.
+ * 服务未连接或终止, 则会立刻启动或重启服务保证可用性.
+ * - 在 [CREATED][Lifecycle.State.CREATED] 和 [STARTED][Lifecycle.State.STARTED] 状态下,
+ * 若服务终止, 不会立刻重启服务, 直到再次进入 [RESUMED][Lifecycle.State.RESUMED] 状态.
+ *
+ * 实现细节:
+ * - 实现 [startService] 方法, 用于实际的启动服务, 并且要连接服务.
+ *
+ * @param startService 启动服务并返回[服务通信对象][T]接口, 若返回 null 代表启动失败.
+ *   这个方法将在 `singleThreadDispatcher` 执行, 并且同时只有一个在执行.
+ * @param singleThreadDispatcher 用于执行内部逻辑的调度器, 需要使用单线程来保证内部逻辑的线程安全.
+ */
+class LifecycleAwareTorrentServiceConnection<T : Any>(
+    parentCoroutineContext: CoroutineContext = EmptyCoroutineContext,
+    singleThreadDispatcher: CoroutineDispatcher,
+    private val lifecycle: Lifecycle,
+    private val startService: suspend () -> T?,
+) : TorrentServiceConnection<T> {
+    private val logger = logger(this::class.simpleName ?: "TorrentServiceConnection")
+
+    // we assert it is a single thread dispatcher
+    private val scope = parentCoroutineContext.childScope(singleThreadDispatcher)
+    
+    private var binderDeferred by atomic(CompletableDeferred<T>())
+
+    private val isAtForeground: MutableStateFlow<Boolean> = MutableStateFlow(false)
+    private val isServiceConnected: MutableStateFlow<Boolean> = MutableStateFlow(false)
+    private val startServiceLock = Mutex()
+
+    override val connected: StateFlow<Boolean> = isServiceConnected
+
+    fun startLifecycleLoop() {
+        scope.launch {
+            lifecycle.repeatOnLifecycle(Lifecycle.State.RESUMED) {
+                isAtForeground.value = true
+                try {
+                    if (!isServiceConnected.value) {
+                        logger.debug { "Lifecycle resume: Service is not connected, start connecting..." }
+                        startServiceWithRetry()
+                    }
+
+                    awaitCancellation()
+                } catch (_: CancellationException) {
+                    isAtForeground.value = false
+                    logger.debug { "Lifecycle pause: App moved to background." }
+                }
+            }
+        }
+    }
+    
+    /**
+     * 服务已断开连接, 通信对象变为不可用.
+     * 如果目前 APP 还在前台, 就要尝试重启并重连服务.
+     */
+    fun onServiceDisconnected() {
+        scope.launch(CoroutineName("TorrentServiceConnection - Lifecycle")) {
+            if (!isServiceConnected.value) {
+                // 已经是断开状态，直接忽略
+                return@launch
+            }
+            isServiceConnected.value = false
```

Comment:
```suggestion
    fun onServiceDisconnected() {
        scope.launch(CoroutineName("TorrentServiceConnection - Lifecycle"), start = UNDISPATCHED) {
            // 这里会是在 onServiceDisconnected 线程, 也就是主线程.
            lock.withLock {
                yield() // 拿到锁了, 可以安全地放开主线程. 进行一次调度, 在 scope 的调度器内继续跑下面代码.
                
                if (!isServiceConnected.value) {
                    // 已经是断开状态，直接忽略
                    return@launch
                }
                isServiceConnected.value = false
```
立即在当前线程启动 job 并且**拿锁**, 保证多线程环境逻辑正确. 

问题场景:
1. 假如 `onServiceDisconnected` 调用后下一瞬间就有 lifecycle RESUME
2. 调度器觉得应该先跑 lifecycle 的 job (非常有可能, 因为它可能更早挂起)
3. lifecycle RESUME 拿到锁, 调用 startService, 完成
4. `onServiceDisconnected` 的 launch 才开始跑, 关掉了本来正常的连接.

在 reason 协程并发安全性问题是, 要考虑调度延迟 - 如果不用 UNDISPATCHED, 协程调度器就可以在任意时间启动它.

Reply:
can we also test this scenario?

Reply:
按照 suggested changes，如果 onServiceDisconnected 和 lifecycle resume 同时执行，那么：

1. onServiceDisconnected 最先被调度并获取锁，释放调度所有权
2. 与此同时 lifecycle resume 执行到 startService 时因无法获取锁而挂起
3. onServiceDisconnected 将 isServiceConnected 设为 false 后释放锁
4. 如果 APP 在前台，lifecycle resume 和 reconnect 同时执行 startService
5. startService 内部的锁又保证了同时只能有一个 starter 在执行，所以不会出现线程安全问题

涉及到需要测试协程调度的 test 有点难写，有参考的例子吗？

Reply:
看看 https://github.com/open-ani/animeko/pull/1616#discussion_r1947831144 能不能解决问题